components/cam/src/physics/cam/check_energy.F90

```diff
@@ -253,107 +255,139 @@ subroutine check_energy_timestep_init(state, tend, pbuf, col_type)
     end if
 
-! Compute vertical integrals of dry static energy and water (vapor, liquid, ice)
+  end subroutine check_energy_timestep_init
+
+!===============================================================================
+  subroutine energy_helper_eam_def(u,v,T,q,ps,pdel,phis, &
+                                   ke,se,wv,wl,wi,wr,ws,te,tw, &     
+                                   ncol,teloc,psterm)
+
+!state vars are of size psetcols,pver, so, not exactly correct
+    real(r8), intent(in) :: u(pcols,pver) 
+    real(r8), intent(in) :: v(pcols,pver) 
+    real(r8), intent(in) :: T(pcols,pver) 
+    real(r8), intent(in) :: q(pcols,pver,pcnst) 
+    real(r8), intent(in) :: ps(pcols) 
+    real(r8), intent(in) :: pdel(pcols,pver) 
+    real(r8), intent(in) :: phis(pcols) 
+
+
+    real(r8), intent(inout) :: ke(ncol)     ! vertical integral of kinetic energy
+    real(r8), intent(inout) :: se(ncol)     ! vertical integral of static energy
+    real(r8), intent(inout) :: wv(ncol)     ! vertical integral of water (vapor)
+    real(r8), intent(inout) :: wl(ncol)     ! vertical integral of water (liquid)
+    real(r8), intent(inout) :: wi(ncol)     ! vertical integral of water (ice)
+    real(r8), intent(inout) :: te(ncol)     ! vertical integral of total energy
+    real(r8), intent(inout) :: tw(ncol)     ! vertical integral of total water
+    real(r8), intent(inout) :: wr(ncol)     ! vertical integral of rain
+    real(r8), intent(inout) :: ws(ncol)     ! vertical integral of snow
+
+    real(r8), intent(inout), optional :: teloc(pcols,pver) 
+    real(r8), intent(inout), optional :: psterm(pcols) 
+
+    integer, intent(in) :: ncol                   
+    integer :: i,k                               
+
     ke = 0._r8
     se = 0._r8
     wv = 0._r8
     wl = 0._r8
     wi = 0._r8
     wr = 0._r8
     ws = 0._r8
-
+#ifdef ENERGY_DIAGNOSTICS
+    if (present(teloc) .and. present(psterm))then
+    teloc = 0.0; psterm = 0.0
+    do k = 1, pver
+       do i = 1, ncol
+          teloc(i,k) = 0.5_r8*(u(i,k)**2 + v(i,k)**2)*pdel(i,k)/gravit &
+                   + t(i,k)*cpair*pdel(i,k)/gravit &
+                   + (latvap+latice)*q(i,k,1       )*pdel(i,k)/gravit
+          if (icldliq > 1  .and.  irain > 1) then
+             teloc(i,k) = teloc(i,k) &
+                      + latice*(q(i,k,icldliq) + q(i,k,irain))*pdel(i,k)/gravit
+          endif
+       end do
+    end do
+    do i = 1, ncol
+       psterm(i) = phis(i)*ps(i)/gravit
+    end do
+    endif
+#endif
     do k = 1, pver
        do i = 1, ncol
-          ke(i) = ke(i) + 0.5_r8*(state%u(i,k)**2 + state%v(i,k)**2)*state%pdel(i,k)/gravit
-          se(i) = se(i) +         state%t(i,k)*cpairv_loc(i,k,lchnk)*state%pdel(i,k)/gravit
-          wv(i) = wv(i) + state%q(i,k,1       )*state%pdel(i,k)/gravit
+          ke(i) = ke(i) + 0.5_r8*(u(i,k)**2 + v(i,k)**2)*pdel(i,k)/gravit
+          se(i) = se(i) + t(i,k)*cpair*pdel(i,k)/gravit
```

Comment:
I may be wrong but from my reading of the code, it looks like cpairv_loc can be different from cpair for WACCM simulations. I haven't run the code in WACM mode to verify that.